Ticket28085 (#38)
torproject/torspec: bandwidth-file-spec.txt

>    1.1.0 - Adds KeyValue Lines to the Header List section, add KeyValues

Reviewer: Start by talking about the most important changes:
```suggestion
   1.1.0 - Add a header containing information about the bandwidth file.
           Document the sbws and Torflow relay line keys.
```

>            to RelayLines and format versions.

Reviewer: ```suggestion
```

>    1.2.0 - Adds new KeyValue Lines to the Header List section with

Reviewer: Start by talking about the most important changes:
```suggestion
   1.2.0 - If there are not enough eligible relays, the bandwidth file
           SHOULD contain a header, but no relays. Adds new
           KeyValue Lines to the Header List section with
```

           statistics about the number of relays included in the file.
           Add new KeyValue Lines to the Relays' Bandwidth List section
           with different bandwidth values.

  All Tor versions can consume format version 1.0.0.

>   All Tor versions can consume format version 1.1.0,

Reviewer: ```suggestion
  All Tor versions can consume format version 1.1.0 and later,
```

    This Line has been added in version 1.1.0 of this specification.

>   "percent_eligible_relays=" Int NL

Reviewer: This spec might be easier to read if you put the keys that don't depend on other keys first.

What do you think of this structure?

Start a new subsection

number_eligible_relays
  * define eligible

minimum_percent_eligible_relays
  * say what happens when we don't have the minimum percentage

number_consensus_relays
percent_eligible_relays
minimum_number_eligible_relays

Author: In this case i should also change what you comment in https://github.com/torproject/torspec/pull/38#discussion_r231435563. I'm fine with that.

    [Zero or one time.]

    The minimum percentage of eligible measured relays with respect
    to the number of relays in the consensus to include in this file.

>     Implementations SHOULD include this KeyValue when

Reviewer: Don't copy and paste the same text into lots of definitions. It will get out of sync.
Instead, start a new subsection, and put the text at the start of the subsection.

>     "minimum_percent_eligible_relays" is minor than

Author: > I think we should always include these keys, so they can be monitored. It also helps parsers if the keys in sbws files are consistent.

ACTION (to myself): change the code to always include these header lines
> 
> If you still want to include this text:
> 
>     * the comparison is the wrong way around
> 
>     * should it depend on the number, not the percentage? 

currently it's implemented as percentage: https://github.com/torproject/sbws/blob/master/sbws/lib/v3bwfile.py#L607. IIRC torflow does also this.

>  Does it matter?

the number of eligible relays is calculated from the percent (https://github.com/torproject/sbws/blob/master/sbws/lib/v3bwfile.py#L834), so no.
> 
> 
> Suggested change Beta
> [Give us feedback](https://www.research.net/r/SuggestedChanges)
>     "minimum_percent_eligible_relays" is minor than
>     "number_eligible_relays" is less than
>     "minimum_number_eligible_relays".

Should i discard then this edit?

>     "minimum_percent_eligible_relays" is minor than

Reviewer: > If you still want to include this text:
> * the comparison is the wrong way around

>     "percent_eligible_relays" and MAY not include any RelayLine.

Author: You would not include <<and MAY not include any RelayLine.>>?. It's what we currently do (though i don't know if it's a good idea)

>     "percent_eligible_relays" and MAY not include any RelayLine.

Reviewer: I do not believe it is a good idea:
> I think we should always include these keys, so they can be monitored. It also helps parsers if the keys in sbws files are consistent.

>     "percent_eligible_relays" and MAY not include any RelayLine.

Author: I understood the sentence about including always the header KeyValues.
What i'm asking is if you think that we should we should remove the sentence ``MAY not include any RelayLine``
Currently, if the there are not enough eligibles relays, *only* the headers are wrotten
(https://trac.torproject.org/projects/tor/ticket/28076)

    This Line has been added in version 1.2.0 of this specification.

  "minimum_number_eligible_relays=" Int NL

    [Zero or one time.]

    The minimum number of relays to include in the bandwidth file.

> 

Reviewer: Include calculations
```suggestion

    This line SHOULD be equal to:
        number_consensus_relays * (minimum_percent_eligible_relays / 100.0)

```

    Implementations SHOULD include this KeyValue when
    "minimum_percent_eligible_relays" is minor than
    "percent_eligible_relays" and MAY not include any RelayLine.

    This Line has been added in version 1.2.0 of this specification.

  "number_consensus_relays=" Int NL

    [Zero or one time.]

    The number of relays in the consensus.

    Implementations SHOULD include this KeyValue when
    "minimum_percent_eligible_relays" is minor than
    "percent_eligible_relays" and MAY not include any RelayLine.

    This Line has been added in version 1.2.0 of this specification.

  "number_eligible_relays=" Int NL

    [Zero or one time.]

>     The number of relays that has been measured and are eligible to

Reviewer: If you're trying to define "eligible", don't use it in the definition.

```suggestion
    The number of relays that have enough measurements to be
    included in the bandwidth file.
```

>     include in the bandwidth file.

Reviewer: ```suggestion
```

Reviewer: If you don't apply this change, then the last line appears twice.

    This Line has been added in version 1.1.0 of this specification.

  "percent_eligible_relays=" Int NL

    [Zero or one time.]

    The number of eligible relays, as a percentage of the number
    of relays in the consensus.
    
    This line SHOULD be equal to:
        (number_eligible_relays * 100.0) / number_consensus_relays
    to the number of relays in the consensus to include in this file.

    Implementations SHOULD include this KeyValue when
    "minimum_percent_eligible_relays" is minor than
    "percent_eligible_relays" and MAY not include any RelayLine.

>     This Line has been added in version 1.2.0 of this specification.

Reviewer: Please add the new format version 1.2.0 in section 1.4:
https://github.com/torproject/torspec/blob/master/bandwidth-file-spec.txt#L43

node_id=$96C15995F30895689291F455587BD94CA427B6FC bw=189 nick=Test2 measured_at=1523911623 updated_at=1523911623 pid_error=3.96703337994 pid_error_sum=3.96703337994 pid_bw=47422125 pid_delta=2.65469736988 circ_fail=0.0 scanner=/filepath

> A.2. Generated by sbws version 0.1.X

Reviewer: ```suggestion
A.2. Generated by sbws version 0.1.0
```

bw=380 error_circ=0 error_misc=0 error_stream=1 master_key_ed25519=YaqV4vbvPYKucElk297eVdNArDz9HtIwUoIeo0+cVIpQ nick=Test node_id=$68A483E05A2ABDCA6DA5A3EF8DB5177638A27F80 rtt=380 success=1 time=2018-05-08T16:13:26
bw=189 error_circ=0 error_misc=0 error_stream=0 master_key_ed25519=a6a+dZadrQBtfSbmQkP7j2ardCmLnm5NJ4ZzkvDxbo0I nick=Test2 node_id=$96C15995F30895689291F455587BD94CA427B6FC rtt=378 success=1 time=2018-05-08T16:13:36

> A.3. Generated by sbws version 0.2.X

Reviewer: Do you mean:
```suggestion
A.3. Generated by sbws version 0.8.1 or later
```

Because the file says:
> software_version=0.8.1-dev0

Author: 0.8.1-dev0 > 0.8.1, 0.8.1 does not generate the new key values

1523911758
version=1.1.0
software=sbws

> software_version=0.1.0

Reviewer: This is not sbws version 0.2.X either.

Author: I forgot to do that change in the code.
The 1st new version that can have all these output correct will be then 1.0.1, since the current version 1.0.0 don't have `version` changed either, i just forgot.

> bw=380 error_circ=0 error_misc=0 error_stream=1 master_key_ed25519=YaqV4vbvPYKucElk297eVdNArDz9HtIwUoIeo0+cVIpQ nick=Test node_id=$68A483E05A2ABDCA6DA5A3EF8DB5177638A27F80 rtt=380 success=1 time=2018-05-08T16:13:26

Reviewer: The terminator above this line is the wrong length.
Is this a bug in sbws?

bw=189 error_circ=0 error_misc=0 error_stream=0 master_key_ed25519=a6a+dZadrQBtfSbmQkP7j2ardCmLnm5NJ4ZzkvDxbo0I nick=Test2 node_id=$96C15995F30895689291F455587BD94CA427B6FC rtt=378 success=1 time=2018-05-08T16:13:36

A.3. Generated by sbws version 1.X.X
[TODO: change the version of sbws that generates this when it is implemented]
1523911758
version=1.2.0
latest_bandwidth=2018-04-16T20:49:18
file_created=2018-04-16T21:49:18
generator_started=2018-04-16T15:13:25
earliest_bandwidth=2018-04-16T15:13:26
num_measured_relays=6000
num_net_relays=6436
num_target_relays=3862
perc_measured_relays=93
perc_measured_targed=60
software=sbws
software_version=1.X.X

> ====

Reviewer: This terminator is the wrong length
```suggestion
=====
```

bw=38000 bw_mean=1127824 bw_median=1180062 desc_avg_bw=1073741824 desc_obs_bw_last=17230879 desc_obs_bw_mean=14732306 error_circ=0 error_misc=0 error_stream=1 master_key_ed25519=YaqV4vbvPYKucElk297eVdNArDz9HtIwUoIeo0+cVIpQ nick=Test node_id=$68A483E05A2ABDCA6DA5A3EF8DB5177638A27F80 rtt=380 success=1 time=2018-05-08T16:13:26
bw=1 bw_mean=199162 bw_median=185675 desc_avg_bw=409600 desc_obs_bw_last=836165 desc_obs_bw_mean=858030 error_circ=0 error_misc=0 error_stream=0 master_key_ed25519=a6a+dZadrQBtfSbmQkP7j2ardCmLnm5NJ4ZzkvDxbo0I nick=Test2 node_id=$96C15995F30895689291F455587BD94CA427B6FC rtt=378 success=1 time=2018-05-08T16:13:36

When there are not enough eligible measured relays:

1540496079
version=1.2.0
earliest_bandwidth=2018-10-20T19:35:52
file_created=2018-10-25T19:35:03
generator_started=2018-10-25T11:42:56
latest_bandwidth=2018-10-25T19:34:39
num_measured_relays=2960
num_net_relays=6436
num_target_relays=3862
perc_measured_relays=46
perc_measured_targed=60
software=sbws

> software_version=1.X.X

Reviewer: The Terminator is missing from the end of the header, and SHOULD be included.
Is this a bug in sbws?
```suggestion
software_version=1.X.X
=====
```

bw=380 error_circ=0 error_misc=0 error_stream=1 master_key_ed25519=YaqV4vbvPYKucElk297eVdNArDz9HtIwUoIeo0+cVIpQ nick=Test node_id=$68A483E05A2ABDCA6DA5A3EF8DB5177638A27F80 rtt=380 success=1 time=2018-05-08T16:13:26
bw=189 error_circ=0 error_misc=0 error_stream=0 master_key_ed25519=a6a+dZadrQBtfSbmQkP7j2ardCmLnm5NJ4ZzkvDxbo0I nick=Test2 node_id=$96C15995F30895689291F455587BD94CA427B6FC rtt=378 success=1 time=2018-05-08T16:13:36

A.3. Generated by sbws version 1.X.X
[TODO: change the version of sbws that generates this when it is implemented]
1523911758
version=1.2.0
software=sbws

> software_version=1.X.X

Reviewer: [TODO: change the version of sbws that generates this when it is implemented]

[TODO: change the version of sbws that generates this when it is implemented]
1523911758
version=1.2.0
latest_bandwidth=2018-04-16T20:49:18
file_created=2018-04-16T21:49:18
generator_started=2018-04-16T15:13:25
earliest_bandwidth=2018-04-16T15:13:26

> num_measured_relays=6000

Reviewer: Please change the names of these keys to match the spec.

Otherwise, metrics will write parsing code using these names, and that code will break when sbws updates to match the spec.

Also change the names of the keys in "When there are not enough eligible measured relays:"

perc_measured_relays=46
perc_measured_targed=60
software=sbws

> software_version=1.X.X

Reviewer: [TODO: change the version of sbws that generates this when it is implemented]

Reviewer: The Terminator is missing from the end of the header, and SHOULD be included.
Is this a bug in sbws?
